File: benchmarks/scripts/dispatch_lookup_bench.mjs
/**
 * Microbenchmark: Python dict vs JS Record (plain object) vs JS Map
 * for operation-dispatch-style lookups.
 *
 * Run:  node benchmarks/scripts/dispatch_lookup_bench.mjs
 *       python3 benchmarks/scripts/dispatch_lookup_bench.py
 */

const ITERS = 10_000_000;
const WARMUP = 1_000_000;

// Build ~400 keys to mirror the real dispatch tables
const keys = Array.from({ length: 400 }, (_, i) => `op_${i}`);
const fns = keys.map((_, i) => () => i);

// --- Record (plain object) ---
const record = Object.create(null);
keys.forEach((k, i) => { record[k] = fns[i]; });

// --- Map ---
const map = new Map();
keys.forEach((k, i) => map.set(k, fns[i]));

// Pick a few keys to cycle through (simulates real workload hitting different ops)
const hotKeys = [keys[0], keys[50], keys[150], keys[300], keys[399]];
const hotLen = hotKeys.length;

function benchRecord() {
  let sum = 0;
  for (let i = 0; i < ITERS; i++) {
    const fn = record[hotKeys[i % hotLen]];
    sum += fn();
  }
  return sum;
}

function benchMap() {
  let sum = 0;
  for (let i = 0; i < ITERS; i++) {
    const fn = map.get(hotKeys[i % hotLen]);
    sum += fn();
  }
  return sum;
}

// --- If/else chain (baseline comparison) ---
function benchIfChain() {
  // Build a chain of 400 comparisons via a generated function
  let sum = 0;
  for (let i = 0; i < ITERS; i++) {
    const key = hotKeys[i % hotLen];
    // Simulate checking a few branches (best/mid/worst case averaged)
    let fn;
    for (let j = 0; j < keys.length; j++) {
      if (key === keys[j]) { fn = fns[j]; break; }
    }
    sum += fn();
  }
  return sum;
}

function run(label, fn) {
  // Warmup
  for (let w = 0; w < WARMUP; w++) {
    const key = hotKeys[w % hotLen];
    record[key]?.();
  }
  fn(); // one full warmup run

  const t0 = performance.now();
  const result = fn();
  const elapsed = performance.now() - t0;
  const opsPerSec = (ITERS / elapsed * 1000).toFixed(0);
  console.log(`${label.padEnd(20)} ${elapsed.toFixed(1).padStart(8)} ms   ${Number(opsPerSec).toLocaleString().padStart(15)} ops/sec   (checksum: ${result})`);
}

console.log(`\nJS dispatch lookup benchmark — ${ITERS.toLocaleString()} iterations, ${keys.length} keys\n`);
run('Record (object)', benchRecord);
run('Map', benchMap);
run('If/else chain', benchIfChain);
console.log();

File: benchmarks/scripts/dispatch_lookup_bench.py
"""
Microbenchmark: Python dict lookup for operation dispatch.

Run:  python3 benchmarks/scripts/dispatch_lookup_bench.py

Compare output with:  node benchmarks/scripts/dispatch_lookup_bench.mjs
"""

import time

ITERS = 10_000_000
WARMUP = 1_000_000

# Build ~400 keys to mirror the real dispatch tables
keys = [f"op_{i}" for i in range(400)]
fns = [lambda i=i: i for i in range(400)]

# Dict dispatch
dispatch = {k: fn for k, fn in zip(keys, fns)}

# Pick a few keys to cycle through
hot_keys = [keys[0], keys[50], keys[150], keys[300], keys[399]]
hot_len = len(hot_keys)


def bench_dict():
    d = dispatch
    hk = hot_keys
    hl = hot_len
    total = 0
    for i in range(ITERS):
        fn = d[hk[i % hl]]
        total += fn()
    return total


def bench_if_chain():
    ks = keys
    fs = fns
    hk = hot_keys
    hl = hot_len
    total = 0
    for i in range(ITERS):
        key = hk[i % hl]
        for j, k in enumerate(ks):
            if key == k:
                total += fs[j]()
                break
    return total


def run(label, fn):
    # Warmup
    fn_warm = dispatch[hot_keys[0]]
    for _ in range(WARMUP):
        fn_warm()

    t0 = time.perf_counter()
    result = fn()
    elapsed = (time.perf_counter() - t0) * 1000  # ms
    ops = int(ITERS / elapsed * 1000)
    print(f"{label:<20} {elapsed:>8.1f} ms   {ops:>15,} ops/sec   (checksum: {result})")


if __name__ == "__main__":
    print(f"\nPython dispatch lookup benchmark — {ITERS:,} iterations, {len(keys)} keys\n")
    run("Dict", bench_dict)
    run("If/elif chain", bench_if_chain)
    print()
